enhancement-proposals/001-security-finding-to-device.md

```diff
@@ -0,0 +1,50 @@
+# Add devices to Security Finding
+
+date: 2023-09-05
+proposer: Jeremy Fisher
+deciders: Kyle, Srot (?)
+
+## Proposed Additions
+
+| Caption       | Name                       | Type       | Description                                      |
+| ------------- | -------------------------- | ---------- | ------------------------------------------------ |
+| Devices       | `security_finding.devices` | `device`   | Allows associating devices to security findings. |
+
+
+## Context
+
+Endpoint protection platforms like CrowdStrike provide information about
+devices implicated in or associated to security findings. OCSF security
+findings can include IP and hostname observables, but OCSF encourages mapping
+"primary" data to named objects. It is rather natural for an analyst to want to
+review identifying information about the endpoint in security findings about
+the endpoint.
+
+This relationship has been discussed in the OCSF Slack workspace; while it
+seems to have buy-in, it hasn't been implemented.
+
+
+## Positive Consequences
+
+We will be able to better present and search on this very relevant information.
+
+## Negative Consequences
+
+OCSF may answer this need differently in the future. This seems unlikely.
+
+
+## Alternatives Considered
+
+We considered mapping IP and hostname to observables, but decided against it
+because: 
```

Comment:
Por qué no los dos? Representing hostnames and IP addresses in observables makes it easy to write queries. By contrast, having a related device makes it possible to represent how the IP/hostname are related to the event, and also allows more details than observable.

Reply:
Why does the query need to change if the same IP is used in both locations?

That reminds me though, the convention we've adopted is observables contain point-in-time snapshots of object state. So if a device had IP address X when the event occurred, this should be represented in observables (same for hostname).

Because we don't have a way to represent snapshots of objects, device records should represent the current state of the device. So the filter `where any(observables, type_id=IP_ADDRESS and value=X)` means you're looking for events where the IP address was X at the time when the event occurred. By contrast, `where any(device, ip=X)` or `any(device, any(network_interfaces, ip=X))` would find events that happened on a device which currently has the IP address X, even if it had a different IP address when the event occurred.

In cases where there's no immutable foreign key from the event to a device, then we probably can't guarantee these semantics. But in the case of CrowdStrike, it has an agent_id attribute that would be used for that, and probably many EDRs have something similar.

Reply:
Shoot. The universe would be easier to model without that pesky time dimension.

This makes me think that what we _really_ want is to return observables in the results, and possibly the non-scalar observables. :(